### PR TITLE
GPG commit validation

### PR DESCRIPTION
**Pull request by sapk**
Display signed commit and validation against gpg key in database. I have started with a very little UI with the a simple title attr indicating the reason. 

Examples:
capture d ecran_2017-03-12_23-29-23

For comparison : https://github.com/sapk/testing-git/commits/master

Need and follow : #710 ✅  go-gitea/git#36 ✅ 

TODO List : 
- [x] UI of commits list
- [x] UI of commit page
- [x] UI file history (same as commits list)
- [x] UI search commit (same as commits list)
- [x] UI diff commit list (same as commits list)
- [x] Add support for translation of new strings

### CODE REVIEW

**Comment by lunny** [comment]
build failed. and it seems the image is not transparent ?

File: models/gpg_key.go
Comment (appleboy) [inline models/gpg_key.go]: please following the style guide. https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md#styleguide

**Review by appleboy — CHANGES_REQUESTED** [review]

File: models/gpg_key.go
Comment (sapk) [reply models/gpg_key.go]: Thx for the catch i will fix it in the PR of the API from where I start this one : #710

**Comment by sapk** [comment]
I didn't precise but this is a really early WIP. 
The build failed is cause by the need of go-gitea/git#36 change.
In this go-gitea/git#36 we need to discuss a litlle about https://github.com/sapk-fork/git-module/blob/1395a17c1a5a05a465032519932794c486e43559/commit.go#L35 and test in this PR if we could do without or if is to much burden and keep the raw payload.

I open this PR mainly to be able to discuss UI design.

**Comment by sapk** [comment]
Validation of commit is ok and display a simple message on hover (title attr) with user/key that succeed or why it failed. (need go-gitea/git#36 to compile)

I need to add more detail in commit page.

For the little API part, I have a problem with json auto-encoding replacing "<" and ">" don't know the good way to tackle this prob in gitea.
https://golang.org/pkg/encoding/json/#Marshal
Example :
```
"verification": {
            "verified": true,
            "reason": "Antoine GIRARD \u003csapk@sapk.fr\u003e / AE8255B8A0D828E4",
            "signature": "-----BEGIN PGP SIGNATURE-----\n\niQEzBAABCAAdFiEEEIOwJg/1vpF1itJ4roJVuKDYKOQFAljC2uIACgkQroJVuKDY\nKOSkQgf+KFT2smNxRlL8S3zptsJAr4fQVV1Lmia5TdepJIUGDvlZMTNuNk2gHqeQ\nnuIJtKGzTD8s5vJIHUQSH2R82oizRPJNSaFKbohTwmMNphQ5uq4gt5DEELkfcNpo\n75N8m0ck+L2+iyPUqmvcofaUqRYjqrvMz/BJ7fB8kEzEJukvjkepfKAQ59T+dzYD\nTd5TY48L18lJv5EdBlNdzpP62/731LEEktUWb6Nttl4MCRXR5CBRPlpBv8rQekjX\nycUAnVM1Yk1BxUQ6zBXJj7xIrc5RChbsDnvAVZmzWtEzME+9w9JiTHg51EJ/fT36\nX6TUL4VQhEhdoQZqJDbzUmIYwPbUoQ==\n=lqpn\n-----END PGP SIGNATURE-----",
            "payload": "tree 7377d79e8a22b4ba522ea204fcdd175490706dd2\nparent 74cde12677bc548487ed1caf192eb19b0b7cf99c\nauthor Antoine GIRARD \u003csapk@sapk.fr\u003e 1488936721 +0100\ncommitter Antoine GIRARD \u003csapk@sapk.fr\u003e 1489165026 +0100\n\nImplement GPG API\n"
        },  
```

**Comment by ghost** [comment]
Great work! Like to see it in Gitea. 👍 
But in my opinion your display solution looks more like a CI indicator. Maybe you want to use something like a lock which is green, open or gray. What do you think?

(And "Message" should be on the same line than all commit messages.)

**Comment by sapk** [comment]
Visual updated in description.

**Comment by ghost** [comment]
@sapk Great! If you only see the question mark, you do not know *what* is missing. I would use a gray lock which indicates that this commit is signed but not trusted.

**Comment by sapk** [comment]
I think that if we use a grey lock it could be understood as trusted (with less security) also by the user and that is not the case. Actually if you hover the question mark it display "No good signature found in keystore" but i plan to find a better describing string like "No known key found for this signature" ?
Furthermore, I will add more detail in the commit page about the signature and is validity against key in db.

**Comment by ghost** [comment]
@sapk I think it will also be much cleaner if you use no border radius between the hash and the symbol. 
noborderradiuslock
One might use `border-top-left-radius` and its variations. :)

**Comment by ghost** [comment]
@sapk I see your point. What about an open gray lock?

**Comment by sapk** [comment]
Changes made 
I also tested changing the icon background a little darker but I don't really know if it is usefull.
capture d ecran_2017-03-13_00-22-05

**Comment by ghost** [comment]
Nice! :100:

Background is ok. Do you think some kind of really small green border around the hash and icon looks good? (Border)

**Comment by sapk** [comment]
Like that ? 
image

**Comment by ghost** [comment]
Yes, looks good. :) Green border: "my hash is trusted" ✔️

**Comment by sapk** [comment]
image
 Better ?

**Comment by ghost** [comment]
This is really good! But then we should add a gray border around untrusted signatures and commits without. :)

**Comment by sapk** [comment]
image

**Comment by sapk** [comment]
Some test for commit page.

image

image

image

**Comment by ghost** [comment]
Looks good. I would use a grey lock on commit page as you have used in the commit list - not a black one.

**Comment by ghost** [comment]
I am not sure if one should also use the border system of the commit list for the whole commit box on these single pages - so a green border around everything, might be too colorful.

**Comment by lunny** [comment]
Yes. green border and background with black text maybe better.

**Comment by ghost** [comment]
@lunny Border and background yes - but I would keep the green font for the signature itself.

**Comment by sapk** [comment]
Like that : 
image
image
Note : now the build should be a succes with the updated git dep

**Comment by ghost** [comment]
Expand the border around "Google GPG" and "Browse Source" but do not change the background. :) I think that was @lunny 's idea.

**Comment by sapk** [comment]
For the border around complete block could be done but i have to tweak a little the css of .attached to change side border on .attached, + bottom for .bottom, + top for .top

I will try something.

**Comment by sapk** [comment]
image
@lunny Like that ?

**Comment by ghost** [comment]
Yes, but not around diffs - I would say.

**Comment by Fastidious** [comment]
I agree with @wevcode, around diffs would be too much, IMHO.

**Comment by lunny** [comment]
@sapk Great and I agree with @wevcode and @Fastidious

**Comment by sapk** [comment]
Ok I will clean this up a little.

**Comment by sapk** [comment]
image

**Comment by Fastidious** [comment]
@sapk excellent, and lovely! This should be added before it breaks with all the UI refactoring happening (and/or about to happen).

**Comment by ghost** [comment]
@sapk Amazing! What is the overall status of this huge feature?

**Comment by sapk** [comment]
I updated the post todo list.
It just need some new translation strings.
For deps of this PR, the git changes are merged and updated in vendor of this PR. 
The #710 could be merge, it "just" need to be finally review and LGTM.

For usability, I talk a little in gitter for the UI to manage users keys. It could be done by api via #710 but really need a easy to use interface. I will go rework a little ssh keys page that could have impact so i will do it in a other PR maybe to be more readable.

**Comment by sapk** [comment]
Translation done this is reviewable now. But since it need #710 maybe wait for it to be merge a just take in account the two last commit.

File: models/error.go
Comment (strk) [inline models/error.go]: May I suggest to put this in error_gpg.go instead ?

**Comment by lunny** [comment]
@sapk #710 merged but there are conflicts

**Comment by sapk** [comment]
I will rebase on current master to not have old commit of GPG api and the squashed one from master to be clean.

File: models/error.go
Comment (sapk) [reply models/error.go]: This is a change maid in #710 but I can split it to make more simple file.

**Comment by ghost** [comment]
Hopefully it will be finished for 1.2.0

File: models/error.go
Comment (strk) [reply models/error.go]: Yes, I meant for the whole GPG block, not just for that single class :)

**Comment by sapk** [comment]
So I rebase on master, this will be easier to review.

**Comment by sapk** [comment]
There is a more efficient way to validate signature that I will have to test more.

**Comment by lunny** [comment]
One could have two same gpg keys?

**Comment by Fastidious** [comment]
@lunny no.

**Comment by sapk** [comment]
@lunny https://github.com/go-gitea/gitea/pull/710/files#diff-212b9fa4c927665ecbc5a5a36382ecedR111

File: models/gpg_key.go
Comment (lunny) [inline models/gpg_key.go]: why use this log?

Comment (sapk) [reply models/gpg_key.go]: That my editor that auto-import :-(

**Comment by sapk** [comment]
So after testing it seems that .IssuerKeyId in sig is not the same the the KeyID of key so remove testing commits.

**Comment by lunny** [comment]
Another thing, maybe all `Id` should be `ID`.

**Comment by sapk** [comment]
@lunny I didn't found any "Id" string in my change where did you see it ?

**Comment by strk** [comment]
@sapk `make lint` should find any of those, but given drone passes I guess there are none ?

**Comment by strk** [comment]
LGTM, great to also see tests in here !

**Comment by strk** [comment]
@sapk I guess migrations are needed, right ?

**Comment by sapk** [comment]
@strk I have try to separate api and db from validation and presentation.
So there is no change in DB in this PR. All change needed were made in the API PR (with a bug for mariadb #1303). All the new types are just for display (CommitVerification, SignCommit). This PR is just validation of commit and display it. API handle change in DB and manipulation via API (Add, remove, list, ...)
A other PR for managing user gpg keys via UI is in WIP. #1293.

I think it will be best to merge it after fix #1303.

**Comment by lunny** [comment]
@sapk #1303 is merged, maybe this need rebase?

**Comment by sapk** [comment]
@lunny I could rebase to be sure that everything works and be more clear but it's not needed. The goal of API #710  was to put the foundation with db format and base object. This PR don't change DB and base object so this would merge seamlessly. All the new code is only call for rendering and don't even write to DB.

**Comment by lunny** [comment]
@sapk this PR is ready to review and merge?

**Comment by sapk** [comment]
Yes and allready have a LGTM ^^

**Comment by strk** [comment]
LGTM , but I think we should deal with vendoring in a different way (as I know another `git` vendor upgrade is needed for the `--follow` support :/)

**Comment by strk** [comment]
Make L-G-T-M work !

**Comment by lunny** [comment]
LGTM

**Pull request merged**
